### ISSUE TITLE
[Feature Request] encrypted send of unencrypted datasets

### ISSUE THREAD

**Issue by c0d3z3r0** [openzfs/zfs#10484]
It would be great to be able to *locally encrypt* plaintext datasets *on the sending side*, so that the receiving side does not have to know the secrets. This is mainly useful for sending backups from LUKS-encrypted systems (ZFS on LUKS) to lesser-trusted systems.

This feature was requested before in #7966 but is closed for unknown reasons (not even the "closed" message is there and posts seem to be missing o.O).

**Comment by johnnyjacq16** [comment]
How to send an encrypted dataset to a non-enecryped dataset, recursively?

**Comment by ahrens** [comment]
@johnnyjacq16 Unless the `--raw`/`-w` flag is specified, zfs send will send decrypted (plaintext) data.  So you can use `zfs send -R ...`.
(Side note, this is kind of the opposite of the feature that this Issue is requesting.)

**Comment by kithrup** [comment]
Pipe the output of zfs send through openssl.  There is no particular way for zfs send, itself, to encrypt an unencrypted dataset (although it would be possible, I suppose, to have the _output_ of zfs send be encrypted, but this would essentially be the same as piping through openssl).

**Comment by misterbigstuff** [comment]
well, no, because piping it into openssl doesn't make it encrypted on the receiving end.

the user is asking for a way to make an encrypted ZFS dataset from an unencrypted one on-the-fly on the sending side, so that the receiver is blind to it.

it would be more like `zfs send -D` that deduplicates streams on-the-fly as they're sent.

i imagine this wouldn't be very maintainable long term, and probably not very useful. if you encrypt data as it sends, do you provide the key every time so the incrementals work? is it just for a full send and not supported by incrementals?

it seems like a mess.

**Comment by c0d3z3r0** [comment]
@kithrup Well, you got it. "There is no particular way for zfs send, itself, to encrypt an unencrypted dataset" <- Thus, this feature request. So what is unclear?

@misterbigstuff

>it would be more like zfs send -D

Exactly!

>maybe just abandon LUKS and use native encryption the whole way through

This is not a solution, when FDE is explicitely wanted. ZFS does not encrypt metadata and does not protect modification (adding, removing volumes) but just encrypts volumes.

>if you encrypt data as it sends, do you provide the key every time so the incrementals work

That would be a requirement, yes. A wrong key would have to be rejected.

>it seems like a mess

If designed carefully, it's not a mess at all. Just because it is impossible to fulfill any wishable use-case, it's not a mess. There are always limitations.

**Comment by misterbigstuff** [comment]
well, then `zfs send -D` is still a good example because of its complexity, it was removed from the codebase.

you can use FDE under ZFS encryption with the associated performance penalty, for the benefit of a much simpler system.

**Comment by c0d3z3r0** [comment]
@misterbigstuff I don't get what you mean by "FDE under ZFS encryption"

**Comment by misterbigstuff** [comment]
well encrypting things on the fly isn't going to be free from a performance standpoint earlier.. it's definitely not free from a code maintenance perspective.

**Comment by c0d3z3r0** [comment]
Yes, that is right. At least it's better to have the performance "problem" only when doing the send instead of at runtime. irt maintenance I can't tell tbh. I had a short look on the `-D` code but I wasn't involved, so I don't know

**Comment by johnnyjacq16** [comment]
@ahrens I tied that but it did not work, I am using the zstd-zfs version of zfs, git clone from BrainSlayer git main or master branch (I taught on github the word master was to be changed to the word main but the main branch is still called master) , been using this version of zfs for the past 6 months having no issues from zstd perspective.

Show below is my setup:

```
[root@Cent_OS-box ~]# lsmod | grep zfs
zfs                  4530176  8
zunicode              335872  1 zfs
zzstd                 565248  1 zfs
zlua                  188416  1 zfs
zcommon                94208  1 zfs
znvpair                98304  2 zfs,zcommon
zavl                   16384  1 zfs
icp                   327680  1 zfs
spl                   118784  6 zfs,icp,zzstd,znvpair,zcommon,zavl
```

```
[root@Cent_OS-box ~]# zfs get compression,encryption,keyformat,keylocation SyS_Backup/ROOT/Arch_OS
NAME                     PROPERTY     VALUE           SOURCE
SyS_Backup/ROOT/Arch_OS  compression  zstd-19         inherited from SyS_Backup
SyS_Backup/ROOT/Arch_OS  encryption   aes-256-gcm     -
SyS_Backup/ROOT/Arch_OS  keyformat    passphrase      -
SyS_Backup/ROOT/Arch_OS  keylocation  prompt          local
[root@Cent_OS-box ~]# 

```

```
zfs send -R SyS_Backup/ROOT/Arch_OS@22.06.2020_Harddrive-Backup | ssh 10.10.10.1 | zfs recv -v ztesting/Test
 && echo 'done!'   
Pseudo-terminal will not be allocated because stdin is not a terminal.
cannot send SyS_Backup/ROOT/Arch_OS@22.06.2020_Harddrive-Backup: encrypted dataset SyS_Backup/ROOT/Arch_OS may 
not be sent with properties without the raw flag
root@10.10.10.1's password:
cannot receive: failed to read from stream
```

-------------------------------------------------------------------------------------------------------------------------------------------------------
Another pool cannot import

Created a disk file use dd if=/dev/zero of=<file name> bs=4k 
Created two partition on the disk file using GTP partition table
         The first small partition contained vfat, and the second partition contained zfs
I imported the pool successfully with the command `zpool import -R <directory> -N  <Original name> <New name>`
Removed some zfs dataset 
Edited the partition table while the pool was still imported to remove the first partition then run partprobe command on the disk file, the partprobe command requested that the pc be rebooted, ran the command `zpool online -e <name of pool> <disk partition>` command then rebooted.

zfs cannot import the pool after reboot.

**Comment by c0d3z3r0** [comment]
@johnnyjacq16 why are you hijacking this feature request for user support?

**Comment by InsanePrawn** [comment]
From my understanding of the openzfs crypto scheme, there is a UX problem with this:
Unless this issue is explicitly addressed in an implementation, I suspect the send stream would generate a new random child key ~~from~~ [cryptographicaly] related to an encryptionroot the user specifies via CLI, right?
The 'random' part of that sentence would by default mean that sending a filesystem or snapshot twice that way (two different zfs send invocations with the exact same args) results in two different, incompatible streams. This is fine for a one-time send only but breaks all kinds of incremental, clonery, etc.
Not sure how the UI to address this would look but it doesn't sound pretty.
(I just realised it might also just create a new encryptionroot, but the same random key problem arises)

AFAIK the current behaviour when specifying a dataset instead of snapshot for plain zfs send is a similar potential UX nightmare for uninformed users where a temporary snapshot is created for the send; I might be wrong though.

**Comment by johnnyjacq16** [comment]
@ahrens @c0d3z3r0 I have now moved my issue case number #10507

**Comment by CySlider** [comment]
I would second this request. As I am also in the same situation right now. I want an offsite backup on an untrusted location and can't shift around my datasets as they are too big. Even if I would be willing to accept the double performance hit of LUKS + ZFS enc.

Can't the "how to store the keys" be solved by maybe also specifying a seed for this use case?

**Comment by csoutreach** [comment]
I also need this feature.
In terms of implementation how about having someway of creating a 'dummy pool/dataset'; with the metadata for the pool/dataset stored in file -- this metadata would include the encryption options (same as normal pool). 

To avoid polluting the existing tools, a new tool e.g. zfspipe could be created. e.g.

So for example:
initially, create a 'dummy' dataset in a dummypool with metadata stored in dummy_pool_meta_store.file

`zfspipe create  -f dummy_pool_meta_store.file -o encryption=aes-128-ccm -o keysource=raw dummypool/dataset`

(with the appropriate selection of the usual creation options from zfs create)

then, anytime the unencrypted data needs to be sent encrypted:

`zfs send realpool/data | zfspipe pipe dummypool/dataset -f dummy_pool_meta_store.file | zfs recv newpool/data`


In terms of implementation, zfspipe would take then stream, receive it as recv would, (and so add encryption) but instead of writing it out to real blocks, it would repackage it into another send stream.

Initial thoughts with regard to maintaining it; main question; 

- would it be easy to receive a stream into this special dummy pool that isn't persistent in the system
- how easy is it for this dummy pool to have a different storage backend, and 
- could this storage backend loopback into using the code to serialize the blocks, without writing to any persistent storage.

I've not looked in full details of the implementation as to how feasible this is, so looking for feedback from someone who knows the codebase well. But thoughts/comments? If it looks feasible, and sensible I'll put the time into working exact details/implementing.

**Comment by c0d3z3r0** [comment]
I like the idea of having a separate tool. Maybe this can be added to zstream? https://openzfs.github.io/openzfs-docs/man/master/8/zstream.8.html
